crates/net/eth-wire/src/muxdemux.rs

```diff
@@ -0,0 +1,422 @@
+            if !tx.is_closed() {
+                return Err(StreamAlreadyExists)
+            }
+        }
+        let (ingress_tx, ingress) = mpsc::unbounded_channel();
+        self.demux.insert(cap, ingress_tx);
+
+        Ok(ingress)
+    }
+
+    fn unmask_msg_id(&self, id: &mut u8) -> Result<CapStreamType, MuxDemuxError> {
+        use SharedCapability::*;
+
+        for cap in self.shared_capabilities.iter_caps() {
+            let offset = cap.relative_message_id_offset();
+            let next_offset = offset + cap.num_messages()?;
+            if *id < next_offset {
+                *id -= offset;
+
+                return match cap {
+                    Eth { .. } => Ok(CapStreamType::new::<EthMessage>()),
+                    UnknownCapability { .. } => Err(CapabilityNotRecognized),
+                }
+            }
+        }
+
+        Err(MessageIdOutOfRange(*id))
+    }
+
+    /// Masks message id with offset relative to the message id suffix reserved for capability
+    /// message ids. The p2p stream further masks the message id (todo: mask whole message id at
+    /// once to avoid copying message to mutate id byte or sink BytesMut).
+    fn mask_msg_id(&self, msg: Bytes) -> Bytes {
+        let mut masked_bytes = BytesMut::zeroed(msg.len());
+        masked_bytes[0] = msg[0] + self.owner.relative_message_id_offset();
+        masked_bytes[1..].copy_from_slice(&msg[1..]);
+
+        masked_bytes.freeze()
+    }
+
+    /// Checks if all clones of this shared stream have been dropped, if true then returns //
+    /// function to drop the stream.
+    fn can_drop(&mut self) -> bool {
+        for tx in self.demux.values() {
+            if !tx.is_closed() {
+                return false
+            }
+        }
+
+        true
+    }
+}
+
+impl<S, E> Stream for MuxDemuxStream<S>
+where
+    S: Stream<Item = Result<BytesMut, E>> + CanDisconnect<Bytes> + Unpin,
+    MuxDemuxError: From<E> + From<<S as Sink<Bytes>>::Error>,
+{
+    type Item = Result<BytesMut, MuxDemuxError>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        // poll once for main stream and each stream clone, since stream clones cannot poll
+        // `MuxDemux`
+        for _ in 0..self.demux.len() + 1 {
+            let res = ready!(self.inner.poll_next_unpin(cx));
+            let mut bytes = match res {
+                Some(Ok(bytes)) => bytes,
+                Some(Err(err)) => return Poll::Ready(Some(Err(err.into()))),
+                None => return Poll::Ready(None),
+            };
+
+            // normalize message id suffix for capability
+            let cap = self.unmask_msg_id(&mut bytes[0])?;
+            if cap == (&self.owner).into() {
+                return Poll::Ready(Some(Ok(bytes)))
+            }
+            let tx = self.demux.get(&cap).ok_or(CapabilityNotConfigured)?;
+            tx.send(bytes).map_err(|_| SendIngressBytesFailed)?;
+        }
+
+        Poll::Ready(None)
```

Comment:
I believe this would terminate the stream if there are more messages on the wire than the number of registered streams?